### PR TITLE
feat(container): add resource containers as service

### PR DESCRIPTION
**Pull request by Monitob**

### CODE REVIEW

File: scaleway/helpers_container.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+	req := ch.setCreateContainerRequest(d)
+	res, err := ch.api.CreateContainer(req, scw.WithContext(ch.ctx))
+	if err != nil {
+		return nil, diag.Errorf("unexpected waiting container error: %s", err)
+	}
+
+	return ch.waitForContainer(res.ID)
+}
+
+func (ch *containerHandler) waitForContainer(containerID string) (*container.Container, diag.Diagnostics) {
```
Comment (remyleone) [inline scaleway/helpers_container.go:118]: Avoid this kind of patterns that make it more difficult to read waiting operation.

File: scaleway/helpers_container.go [hunk 2]
```diff
@@ -35,3 +35,65 @@ func containerAPIWithRegionAndID(m interface{}, id string) (*container.API, scw.
...
 	return api, region, id, nil
 }
+
+func setCreateContainerRequest(d *schema.ResourceData, region scw.Region) (*container.CreateContainerRequest, error) {
+	// required
+	nameRaw := d.Get("name")
+	namespaceID := d.Get("namespace_id")
+
+	name := expandOrGenerateString(nameRaw.(string), "co")
+	privacyType := d.Get("privacy") // default unknown_privacy
```
Comment (remyleone) [inline scaleway/helpers_container.go:45]: Why do you specify unknown_privacy? It is something that is purely internal and should not be used by client. Default privacy is private and/or public but not unknown.
Protocol default is probably http1 but not unknown.

File: scaleway/resource_container_test.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+					resource.TestCheckResourceAttr("scaleway_container.main", "timeout", "300"),
+					resource.TestCheckResourceAttr("scaleway_container.main", "max_concurrency", "50"),
+					resource.TestCheckResourceAttr("scaleway_container.main", "redeploy", "false"),
+					resource.TestCheckResourceAttr("scaleway_container.main", "privacy", container.ContainerPrivacyPublic.String()),
+					resource.TestCheckResourceAttr("scaleway_container.main", "protocol", container.ContainerProtocolHTTP1.String()),
+				),
+			},
+			{
+				Config: `
+					resource scaleway_container_namespace main {
```
Comment (remyleone) [inline scaleway/resource_container_test.go:86]: What is the difference between this step and the step before?

File: scaleway/resource_container_test.go [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+					testRemoveAllRegistryNamespaceGenerated(tt),
+				),
+			},
+		},
+	})
+}
+
+func TestAccScalewayContainer_WithIMG(t *testing.T) {
+	if !*UpdateCassettes {
+		t.Skip("Skipping Container test as with registry image")
```
Comment (remyleone) [inline scaleway/resource_container_test.go:105]: Add more explanation. Such as: We cannot capture the traffic emitted by docker when pushing to the registry. Therefore we cannot re run the complete e2e with cassettes. Or something like that.

File: scaleway/resource_container.go [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+	res, err := api.CreateContainer(req, scw.WithContext(ctx))
+	if err != nil {
+		return diag.Errorf("creation container error: %s", err)
+	}
+
+	d.SetId(newRegionalIDString(region, res.ID))
+
+	// verify name space state
+	_, err = api.WaitForNamespace(&container.WaitForNamespaceRequest{
```
Comment (remyleone) [inline scaleway/resource_container.go:195]: Should we wait for the Container AND the Namespace ?

File: scaleway/resource_container.go [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+	return resourceScalewayContainerRead(ctx, d, meta)
+}
+
+func resourceScalewayContainerRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api, region, containerID, err := containerAPIWithRegionAndID(meta, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	co, err := api.WaitForContainer(&container.WaitForContainerRequest{ContainerID: containerID,
```
Comment (remyleone) [inline scaleway/resource_container.go:212]:
```suggestion
	co, err := api.WaitForContainer(&container.WaitForContainerRequest{
	ContainerID: containerID,
```

Comment (remyleone) [reply scaleway/resource_container.go:212]: use c instead of co. It is smaller.

File: scaleway/resource_container.go [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+	_, err = api.WaitForNamespace(&container.WaitForNamespaceRequest{
+		NamespaceID: expandID(namespaceID),
+		Region:      region,
+	}, scw.WithContext(ctx))
+	if err != nil {
+		return diag.Errorf("unexpected namespace error: %s", err)
+	}
+
+	// check for container state
+	_, err = api.WaitForContainer(&container.WaitForContainerRequest{ContainerID: containerID,
```
Comment (remyleone) [inline scaleway/resource_container.go:260]:
```suggestion
	_, err = api.WaitForContainer(&container.WaitForContainerRequest{
	ContainerID: containerID,
```

File: scaleway/resource_container.go [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+
+	if d.HasChanges("port") {
+		req.Port = toUint32(d.Get("port"))
+	}
+
+	if d.HasChanges("redeploy") {
+		req.Redeploy = expandBoolPtr(d.Get("redeploy"))
+	}
+
+	_, err = api.UpdateContainer(
```
Comment (remyleone) [inline scaleway/resource_container.go:356]: Feel free to put it all on a single line in that case.

File: scaleway/resource_container_test.go [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+		encodedJSON, err := json.Marshal(authConfig)
+		if err != nil {
+			return fmt.Errorf("could not marshal auth config: %v", err)
+		}
+
+		ctx := context.Background()
+		authStr := base64.URLEncoding.EncodeToString(encodedJSON)
+
+		out, err := cli.ImagePull(ctx, "docker.io/library/alpine", types.ImagePullOptions{})
```
Comment (remyleone) [inline scaleway/resource_container_test.go:313]: Could you introduce a variable for this image? Maybe we want to change it later.

File: docs/data-sources/container.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    registry_image = "${scaleway_container_namespace.main.endpoint}/alpine:test"
+    port = 9997
+    cpu_limit = 140
+    memory_limit = 256
+    min_scale = 3
+    max_scale = 5
+    timeout = 600
+    max_concurrency = 80
+    privacy = "private"
+    protocol = "h2c"
```
[the remark this answers is not in the document]
Comment (Monitob) [reply docs/data-sources/container.md:37]: It's the cleartext variant for HTTP/2 protocol. Please check [this](https://httpd.apache.org/docs/2.4/howto/http2.html). I'm gonna improve the documentation for more details about the protocols.

File: scaleway/resource_container.go (hunk 5 above)
Comment (Monitob) [reply scaleway/resource_container.go:195]: This is the normal check when for resource dependency. I'm gonna try the test if it is really necessary.

File: scaleway/resource_container_test.go (hunk 3 above)
Comment (Monitob) [reply scaleway/resource_container_test.go:86]: This was added to force the `registry_namespace` created as default. I'm gonna remove this and the namespaces are going to be treated by the sweeper process.

File: scaleway/helpers.go [hunk 11]
```diff
@@ -503,6 +503,30 @@ func validateDuration() schema.SchemaValidateFunc {
 	}
 }
 
+func validateMaxLimit(max uint64) schema.SchemaValidateDiagFunc {
```
[the remark this answers is not in the document]
Comment (Monitob) [reply scaleway/helpers.go:506]: The signature `SchemaValidateFunc` is now deprecated, I was trying the new signature. I'll roll back with the standard validator lib by terraform.

File: docs/data-sources/container.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+- `name` - (Required) The unique name of the container namespace.
+
+~> **Important** Updates to `name` will recreate the container.
+
+- `description` (Optional) The description of the container.
+
+- `namespace_id` - (Required) The container namespace ID of the container.
+
+- `region` - (Defaults to [provider](../index.md#region) `region`). The [region](../guides/regions_and_zones.md#regions) in which the namespace should be created.
```
[the remark this answers is not in the document]
Comment (Monitob) [reply docs/data-sources/container.md]: The attribute `region` will be read-only.

File: docs/data-sources/container.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Monitob) [reply docs/data-sources/container.md:37]: I've added a new protocol [section](https://github.com/scaleway/terraform-provider-scaleway/pull/1103/files#diff-79bdf83c01d710974aa062a5592c0f039b32607bde22c3be1101839e7b2cc5e9R104)

File: docs/data-sources/container.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+- `environment_variables` - The environment variables of the container.
+
+- `min_scale` - The minimum of running container instances continuously. Defaults to 0.
+
+- `max_scale` - The maximum of number of instances this container can scale to. Default to 20.
+
+- `memory_limit` - The memory computing resources in MB to allocate to each container. Defaults to 128.
+
+- `cpu_limit` - The amount of vCPU computing resources to allocate to each container. Defaults  to 70.
```
[the remark this answers is not in the document]
Comment (Monitob) [reply docs/data-sources/container.md:72]: I've just added an [explanation](https://github.com/scaleway/terraform-provider-scaleway/pull/1103/files#diff-79bdf83c01d710974aa062a5592c0f039b32607bde22c3be1101839e7b2cc5e9R118).

File: scaleway/resource_container.go [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+			},
+			"protocol": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The communication protocol. Defaults to http1.",
+				Default:     container.ContainerProtocolHTTP1.String(),
+				ValidateFunc: validation.StringInSlice([]string{
+					container.ContainerProtocolH2c.String(),
+					container.ContainerProtocolHTTP1.String(),
+					container.ContainerProtocolUnknownProtocol.String()},
```
[the remark this answers is not in the document]
Comment (Monitob) [reply scaleway/resource_container.go]: Type unknown is no longer considered.

File: scaleway/data_source_container_test.go [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+						container_id = scaleway_container.main.id
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayContainerExists(tt, "scaleway_container.main"),
+					resource.TestCheckResourceAttr("scaleway_container.main", "name", "test-container-data"),
+					resource.TestCheckResourceAttrSet("data.scaleway_container.by_name", "id"),
+
+					resource.TestCheckResourceAttr("data.scaleway_container.by_id", "name", "test-container-data"),
+					resource.TestCheckResourceAttrSet("data.scaleway_container.by_id", "id"),
```
[the remark this answers is not in the document]
Comment (Monitob) [reply scaleway/data_source_container_test.go:45]: Added

File: scaleway/helpers_container.go (hunk 2 above)
Comment (Monitob) [reply scaleway/helpers_container.go]: type unknown is no longer considered

File: docs/resources/container.md [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+The vCPU represents a portion or share of the underlying, physical CPU that is assigned to a particular virtual machine (VM).
+
+You may decide how much computing resources to allocate to each container.
+The `memory_limit` (in MB) must correspond with the right amount of vCPU.
+
+**Important:** The right choice for your container's resources is very important, as you will be billed based on compute usage over time and the number of Containers executions.
+
+Please check our [price](https://www.scaleway.com/en/docs/faq/serverless-containers/#prices) section for more details.
+
+| Memory (in MB)      | vCPU |
```
Comment (remyleone) [inline docs/resources/container.md:127]: Use http://markdowntable.com/ to have a nice formatted markdown table :)

File: docs/data-sources/container.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
+---
+page_title: "Scaleway: scaleway_container"
+description: |-
+Gets information about a container.
+---
+
+# scaleway_container
+
+Creates and manages Scaleway Container.
```
Comment (remyleone) [inline docs/data-sources/container.md:9]: If it is a data source it should not create anything...

File: docs/resources/container.md [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+resource scaleway_container_namespace main {
+}
+
+resource scaleway_container main {
+    name = "test-container-data"
+    namespace_id = scaleway_container_namespace.main.id
+}
+
+// Get info by container name
+data "scaleway_container" "by_name" {
```
Comment (remyleone) [inline docs/resources/container.md:29]: This should go in the documentation of the data source.

File: docs/resources/container.md [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+## Privacy
+
+By default, creating a container will make it `public`, meaning that anybody knowing the endpoint could execute it.
+A container can be made `private` with the privacy parameter.
+
+Please check our [authentication](https://developers.scaleway.com/en/products/containers/api/#protocol-9dd4c8) section
+
+## Memory and vCPUs configuration
+
+The vCPU represents a portion or share of the underlying, physical CPU that is assigned to a particular virtual machine (VM).
```
Comment (remyleone) [inline docs/resources/container.md:118]: How is this portion measured? What is the maximum vCPU possible in theory? Is it a gauge?

File: docs/resources/container.md [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+Please check our [price](https://www.scaleway.com/en/docs/faq/serverless-containers/#prices) section for more details.
+
+| Memory (in MB)      | vCPU |
+| ----------- | ----------- |
+| 128      | 70m       |
+| 256      | 140m       |
+| 512      | 280m       |
+| 1024      | 560m       |
+
+**Note:** 560mCPU accounts roughly for half of one CPU power of a Scaleway General Purpose instance
```
[the remark this answers is not in the document]
Comment (Monitob) [reply docs/resources/container.md:134]: Phrase from [here](https://developers.scaleway.com/en/products/containers/api/#containers). It refers to the last configuration case.

**Review by remyleone — APPROVED** [review]

**Pull request merged**